ocaml/xapi/xapi_session.ml

```diff
@@ -442,31 +441,49 @@ let revalidate_external_session ~__context ~session =
             (trackid session)
         in
         debug "%s" msg ;
-        destroy_db_session ~__context ~self:session
+        destroy_db_session ~__context ~self:session ;
+        acc
       ) else
         (* otherwise, we try to revalidate it against the external authentication service *)
         let session_lifespan = 60.0 *. 30.0 in
         (* allowed session lifespan = 30 minutes *)
         let random_lifespan = Random.float 60.0 *. 10.0 in
 
-        (* extra random (up to 10min) lifespan to spread access to external directory *)
-
         (* 2. has the external session expired/does it need revalidation? *)
         let session_last_validation_time =
           Date.to_unix_time
             (Db.Session.get_validation_time ~__context ~self:session)
         in
         let now = Date.now () in
-        let session_needs_revalidation =
+        let session_timeout =
           Date.to_unix_time now
           > session_last_validation_time +. session_lifespan +. random_lifespan
         in
-        if session_needs_revalidation then (
+
+        (* extra random (up to 10min) lifespan to spread access to external directory *)
+        let authenticated_user_sid =
+          Db.Session.get_auth_user_sid ~__context ~self:session
+        in
+        let user_validated =
+          (* acc is [(sid, check_result)] , true for check pass, false for check failed *)
+          match List.assoc_opt authenticated_user_sid acc with
```

Comment:
a StringMap (Map.Make(String)) would be better than a list. O(log n) lookup, instead of O(n).

Reply:
In my first version I do use it. 
but I do not expect the list is large as we usually have very little users/subject been added into XenServer. 
Then I chose to use List for code simplify. .